README.md



# geoChina

[![Build Status](https://travis-ci.org/caijun/geoChina.png?branch=master)](https://travis-ci.org/caijun/geoChina) [![Issue Stats](http://issuestats.com/github/caijun/geoChina/badge/issue?style=flat)](http://issuestats.com/github/caijun/geoChina)

An R package for geocoding address by calling the Google or Baidu Geocoding API and coordinate conversions for WGS-84, GCJ-02 and BD-09 geodetic system

## Introduction

In China the digital maps use two main Geographic Coordinate Systems (GCSs), namely GCJ-02 and BD-09, rather than the well known WGS-84 because of information security. geoChina provides functions for the mutual coordinate conversions between those three GCSs. Therefore, when geocoding address or reverse geocoding latitude/longitude location by calling either Google or Baidu Geocoding API, geoChina allows for inputting or outputting the coordinates in the GCS different from the one that Google or Baidu map uses. Besides, function for geocoding IP address is provided.

## Installation

Install the latest development version from GitHub (`devtools` package is required):

```r
devtools::install_github("caijun/geoChina")
```

## Contact

Bugs and feature requests can be filed to
<https://github.com/caijun/geoChina/issues>. Pull requests are also welcome.


NEWS
Version 1.2.7
-------------------------------------------------------------------------
NEW FEATURES

* geocode() can provide lat/lng coordinates with formatted address and adress components using output = 'latlnga'.



Version 1.2.6
-------------------------------------------------------------------------
NEW FEATURES

* add some tests

FIXS

* when geocoding address with space, gencode returns HTTP error 400, which is fixed by encode URL string



Version 1.2.5
-------------------------------------------------------------------------
NEW FEATURES

* use curl supported by libcurl other than url to open URLs
* remove URLencode to show URLs human friendly



Version 1.2.4
-------------------------------------------------------------------------
NEW FEATURES

* add time parameter to specify time interval for (rev)geocode to avoid exceeding usage limits of Google Maps API



Version 1.2.3
-------------------------------------------------------------------------
NEW FEATURES

* geohost uses freegeoip.net or ipinfo.io rather than hostip.info IP lookup API
* store IP country to avoid calling geohost repeatedly when (rev)geocoding muptiple addresses or locations



Version 1.2.2
-------------------------------------------------------------------------
FIXS

* geocode and revgeocode failed to call google api when user's location can't be geolocated by IP on Windows



Version 1.2.1
-------------------------------------------------------------------------
FIXS

* geocode and revgeocode failed to call google api when user's location can't be geolocated by IP



Version 1.2
-------------------------------------------------------------------------
NEW FEATURES

* add geohost function for IP address lookup.

DOCUMENTATION CHANGES

* minor revision.



Version 1.1.3
-------------------------------------------------------------------------
NEW FEATURES

* automaticaly specify different google maps api urls according to user's location geolocated by IP. For users inside China, ditu.google.cn api is used; otherwise, maps.google.com api is used.

FIXS

* geocode and revgeocode failed with Error "unable to connect to 'maps.googleapis.com' on port 80"



Version 1.1.2
-------------------------------------------------------------------------
NEW FEATURES

* minor change to geocode/revgeocode to make function optimized and using friendly .

FIXS

* geocode address out of China with no results returned by calling google map api

DOCUMENTATION CHANGES

* more detail on ics and ocs parameters.



Version 1.1.1
-------------------------------------------------------------------------
FIXS

* revgeocode failed with Error "could not find function "NULLtoNA"" by calling baidu map api
* minor spelling error in DESCRIPTION



Version 1.1
-------------------------------------------------------------------------
NEW FEATURES

* all functions support vectorize operation.

FIXS

* geocode Chinese address failed with status INVALID_REQUEST by calling google map api on Windows platform.
* geocode with turning messaging on failed with Error "invalid multibyte string 1" by calling google map api on Windows platform.

DOCUMENTATION CHANGES

* add examples to illustrate vectorize operation feature.



Version 1.0
-------------------------------------------------------------------------
PACKAGE GENESIS.


tests/testthat.R
library(testthat)

test_check("geoChina")


tests/testthat/test-geocode.R
context("geocode")

test_that("geocode with google api", {
  coords <- geocode('Tsinghua University', api = 'google')
  expect_true(abs(coords$lat - 39.9984) < 5e-3)
  expect_true(abs(coords$lng - 116.3203) < 5e-3)
})

R/conv.R
#' Convert coordinates
#'
#' the general function that converts lat/lon coordintes from one GCS to another 
#' GCS including WGS-84, GCJ-02 and BD-09 either locally or by calling Baidu 
#' Maps API.
#' 
#' @param lat a numeric latitude
#' @param lon a numeric longitude
#' @param from the inputting GCS
#' @param to the outputting GCS
#' @param api use baidu maps api. Note that baidu maps api only supports the 
#' transformations from WGS-84 or GCJ-02 to BD-09. Other coodinate conversions 
#' must be done locally. As the conversion result is the same, it's recommended 
#' to perform conversions locally.
#' @return a data.frame with variables lat/lng 
#' @author Jun Cai (\email{cai-j12@@mails.tsinghua.edu.cn}), PhD student from 
#' Center for Earth System Science, Tsinghua University
#' @details note that the baidu maps api limits to 20 lat/lon coordinates per query. 
#' Since the coordinate conversion results of Baidu Maps API and local algorithms 
#' are the same, it is recommended to use local algorithms.
#' @seealso \code{\link{wgs2gcj}}, \code{\link{wgs2bd}}, \code{\link{gcj2wgs}}, 
#' \code{\link{gcj2bd}}, \code{\link{bd2wgs}}, \code{\link{bd2gcj}}.
#' @export
#' @import curl RJSONIO
#' @examples
#' \dontrun{
#' # latitude/longitude coordinates of Beijing railway station
#' # WGS-84: (39.90105, 116.42079)
#' # GCJ-02: (39.90245, 116.42703)
#' # BD-09:  (39.90851, 116.43351)
#' conv(39.90105, 116.42079, from = 'WGS-84', to = 'GCJ-02')
#' conv(39.90105, 116.42079, from = 'WGS-84', to = 'GCJ-02', api = TRUE)
#' conv(39.90105, 116.42079, from = 'WGS-84', to = 'BD-09')
#' conv(39.90105, 116.42079, from = 'WGS-84', to = 'BD-09', api = TRUE)
#' conv(39.90245, 116.42703, from = 'GCJ-02', to = 'WGS-84')
#' # not supported by baidu map api, return NAs
#' conv(39.90245, 116.42703, from = 'GCJ-02', to = 'WGS-84', api = TRUE)
#' conv(39.90245, 116.42703, from = 'GCJ-02', to = 'BD-09')
#' conv(39.90245, 116.42703, from = 'GCJ-02', to = 'BD-09', api = TRUE)
#' conv(39.90851, 116.43351, from = 'BD-09', to = 'GCJ-02')
#' # not supported by baidu map api, return NAs
#' conv(39.90851, 116.43351, from = 'BD-09', to = 'GCJ-02', api = TRUE)
#' conv(39.90851, 116.43351, from = 'BD-09', to = 'WGS-84')
#' # not supported by baidu map api, return NAs
#' conv(39.90851, 116.43351, from = 'BD-09', to = 'WGS-84', api = TRUE)
#' # convert multiple coordinates
#' lat = c(39.99837, 39.98565)
#' lng = c(116.3203, 116.2998)
#' conv(lat, lng, from = 'WGS-84', to = 'GCJ-02')
#' }

conv <- function(lat, lon, from = c('WGS-84', 'GCJ-02', 'BD-09'), 
                 to = c('WGS-84', 'GCJ-02', 'BD-09'), api = FALSE){  
  # check parameters
  stopifnot(is.numeric(lat))
  stopifnot(is.numeric(lon))
  from <- match.arg(from)
  to <- match.arg(to)
  stopifnot(is.logical(api))
  
  # vectorize
  if(length(lat) > 1){
    return(ldply(seq_along(lat), function(i){conv(lat[i], lon[i], from = from, 
                                                  to = to, api = api) }))
  }
  
  if(from == to){
    return(data.frame(lat = lat, lng = lon))
  } else{
    if(api){
      # coordinate system lookup table
      code <- c(0, 2, 4)
      names(code) <- c('WGS-84', 'GCJ-02', 'BD-09')
      f <- code[from]
      t <- code[to]
      
      # format url
      # http://api.map.baidu.com/ag/coord/convert?x=lon&y=lat&from=FROM&to=TO
      url_string <- paste('http://api.map.baidu.com/ag/coord/convert?x=', lon, 
                          '&y=', lat, '&from=', f, '&to=', t, sep = '')
      message(paste('calling ', url_string, ' ... ', sep = ''), appendLF = F)
      
      # convert
      con <- curl(URLencode(url_string))
      cv <- fromJSON(paste(readLines(con, warn = FALSE), collapse = ''), drop = FALSE)
      message('done.')  
      close(con)
      
      # did convert fail?
      if(is.list(cv)){
        if(cv$error == 0){
          cvdf <- with(cv, {data.frame(lat = as.numeric(base64(y, FALSE)), 
                                       lng = as.numeric(base64(x, FALSE)), 
                                       row.names = NULL)})
          return(cvdf)
        }
      } else{
        warning(paste('convert failed with error ', cv['error'], 
                      '. conversion is not supported by baidu map api.',
                      sep = ''), 
                call. = FALSE)
        return(data.frame(lat = NA, lng = NA))
      }
    } else{
      if(from == 'WGS-84' & to == 'GCJ-02') return(wgs2gcj(lat, lon))
      if(from == 'WGS-84' & to == 'BD-09') return(wgs2bd(lat, lon))
      if(from == 'GCJ-02' & to == 'WGS-84') return(gcj2wgs(lat, lon))
      if(from == 'GCJ-02' & to == 'BD-09') return(gcj2bd(lat, lon))
      if(from == 'BD-09' & to == 'WGS-84') return(bd2wgs(lat, lon))
      if(from == 'BD-09' & to == 'GCJ-02') return(bd2gcj(lat, lon))
    }
  }
}

R/geohost.R
#' IP address lookup
#'
#' geocodes an IP address using either freegeoip.net \url{http://freegeoip.net} or 
#' ipinfo.io \url{http://ipinfo.io/developers} IP lookup API.
#' 
#' @param ip a character vector specifying an IP (e.g., "12.215.42.19"). 
#' The default value is no IP is specified and the host IP is used.
#' @param api use freegeoip.net or ipinfo.io IP lookup API. By default 
#' freegeoip.net is used.
#' @return a vector with information of ip, country_code, country_name, region_code, 
#' city, zip_code, time_zone, latitude, longitude and metro_code for freegeoip.net API, 
#' of ip, hostname, city, region, country, loc, org, postal and phone for ipinfo.io 
#' IP lookup API. If numerous IPs are inputted, a data.frame is returned. 
#' @author Jun Cai (\email{cai-j12@@mails.tsinghua.edu.cn}), PhD student from 
#' Center for Earth System Science, Tsinghua University
#' @details note that freegeoip.net API is allowed up to 10,000 queries per hour 
#' by default, ipinfo API is limited to 1,000 requests per day.
#' @seealso \code{\link{geocode}}, \code{\link{revgeocode}}.
#' 
#' freegeoip.net IP lookup API at \url{http://freegeoip.net}
#' 
#' ipinfo.io IP lookup API at \url{http://ipinfo.io/developers}
#' @export
#' @examples
#' \dontrun{
#' # geocode host IP
#' geohost()
#' geohost(api = "ipinfo.io")
#' # specify an IP for geocoding
#' geohost(ip = "12.215.42.19")
#' geohost(ip = "12.215.42.19", api = "ipinfo.io")
#' # geocode multiple IPs
#' geohost(ip = c("61.135.169.81", "12.215.42.19"))
#' geohost(ip = c("61.135.169.81", "12.215.42.19"), api = "ipinfo.io")
#' }

geohost <- function(ip = '', api = c("freegeoip.net", "ipinfo.io")) {
  # check parameters
  stopifnot(is.character(ip))
  api <- match.arg(api)
  
  # vectorize for many IPs
  if (length(ip) > 1) {
    return(ldply(ip, geohost, api = api))
  }
  
  if (api == "freegeoip.net") {
    url_string <- "http://freegeoip.net/json/"
    if (nchar(ip) > 0) {
      url_string <- paste0(url_string, ip)
    }
    con <- curl(url_string)
    x <- readLines(con, warn = F)
    close(con)
    as.data.frame(fromJSON(x))
  } else {
    url_string <- "http://ipinfo.io/json"
    if (nchar(ip) > 0) {
      url_string <- paste0("http://ipinfo.io/", ip, "/json")
    }
    con <- curl(URLencode(url_string))
    x <- paste(readLines(con, warn = F), collapse = "")
    close(con)
    as.data.frame(as.list(fromJSON(x)))
  }
}

R/revgeocode.R
#' Reverse geocode
#'
#' reverse geocodes a lat/lng location using Google or Baidu Maps API.  Note that in 
#' most cases by using this function you are agreeing to the Google Maps API Terms 
#' of Service at \url{https://developers.google.com/maps/terms} or the Baidu Maps 
#' API Terms of Use at \url{http://developer.baidu.com/map/law.htm}.
#' 
#' @param latlng a location in latitude/longitude format
#' @param ics the coordinate system of inputing location, including WGS-84, GCJ-02 
#' and BD-09, which are the GCSs of Google Earth, Google Map in China and Baidu 
#' Map, respectively. For location out of China, ics is automatically set to 'WGS-84'
#' and other values are ignored.
#' @param api use google or baidu maps api
#' @param key an api key must be provided when calling baidu maps api. 
#' While it's unnecessary for calling google maps api.
#' @param output formatted address or formmatted address with address components
#' @param messaging turn messaging on/off. The default value is FALSE.
#' @param time the time interval to revgeocode, in seconds. Default value is zero. 
#' When you revgeocode multiple locations, set a proper time interval to avoid 
#' exceeding usage limits. For details see 
#' \url{https://developers.google.com/maps/documentation/business/articles/usage_limits}
#' @return a data.frame with variables address or detail address components 
#' @author Jun Cai (\email{cai-j12@@mails.tsinghua.edu.cn}), PhD student from 
#' Center for Earth System Science, Tsinghua University
#' @details note that the google maps api limits to 2500 queries a day.
#' @seealso \code{\link{geocode}}, \code{\link{geohost}}.
#' 
#' Google Maps API at \url{http://code.google.com/apis/maps/documentation/geocoding/} 
#' and Baidu Maps API at \url{http://developer.baidu.com/map/webservice-geocoding.htm}
#' @export
#' @examples
#' \dontrun{
#' # reverse geocode Beijing railway station
#' revgeocode(c(39.90105, 116.42079), ics = 'WGS-84', api = 'google', 
#'            output = 'address')
#' revgeocode(c(39.90245, 116.42703), ics = 'GCJ-02', api = 'google', 
#'            output = 'address', messaging = TRUE)
#' revgeocode(c(39.90851, 116.43351), ics = 'BD-09', api = 'google', 
#'            output = 'addressc')
#' revgeocode(c(39.90851, 116.43351), ics = 'BD-09', api = 'baidu', 
#'            key = 'your baidu maps api key', output = 'address')
#' revgeocode(c(39.90245, 116.42703), ics = 'GCJ-02', api = 'baidu', 
#'            key = 'your baidu maps api key', output = 'address', messaging = TRUE)
#' revgeocode(c(39.90105, 116.42079), ics = 'WGS-84', api = 'baidu', 
#'            key = 'your baidu maps api key', output = 'addressc')
#' # reverse geocode multiple locations
#' latlng = data.frame(lat = c(39.99837, 39.98565), lng = c(116.3203, 116.2998))
#' revgeocode(latlng, ics = 'WGS-84', api = 'google', output = 'address')
#' revgeocode(latlng, ics = 'WGS-84', api = 'google', output = 'address', time = 2)
#' }

revgeocode <- function(latlng, ics = c('WGS-84', 'GCJ-02', 'BD-09'), 
                       api = c('google', 'baidu'), key = '', 
                       output = c('address', 'addressc'), messaging = FALSE, 
                       time = 0){
  # check parameters
  stopifnot(class(latlng) %in% c('numeric', 'data.frame'))
  ics <- match.arg(ics)
  api <- match.arg(api)
  stopifnot(is.character(key))
  output <- match.arg(output)
  stopifnot(is.logical(messaging))
  stopifnot(is.numeric(time))
  
  # vectorize for many locations
  if(is.data.frame(latlng)){
    return(ldply(seq_along(latlng), function(i){ 
      Sys.sleep(time)
      revgeocode(as.numeric(latlng[i, ]), ics = ics, api = api, key = key, 
                 output = output, messaging = messaging) }))
  }
  
  # different google maps api is used based user's location. If user is inside China,
  # ditu.google.cn is used; otherwise maps.google.com is used.
  if(api == 'google'){
    cname <- ip.country()
    if(cname == "CN"){
      api_url <- 'http://ditu.google.cn/maps/api/geocode/json'
    } else{
      api_url <- 'http://maps.googleapis.com/maps/api/geocode/json'
    }
  } else{
    api_url <- 'http://api.map.baidu.com/geocoder/v2/'
  }
  
  # format url
  if(api == 'google'){
    # convert coordinates only in China
    if(!outofChina(latlng[1], latlng[2])){
      latlng <- conv(latlng[1], latlng[2], from = ics, to = 'GCJ-02')
    } else{
      if(ics != 'WGS-84'){
        message('wrong usage: for location out of China, ics can only be set to "WGS-84"', appendLF = T)
      }
    }
    
    # http://maps.googleapis.com/maps/api/geocode/json?latlng=LAT,LNG
    # &sensor=FALSE&key=API_KEY for outside China
    # http://ditu.google.com/maps/api/geocode/json?latlng=LAT,LNG
    # &sensor=FALSE&key=API_KEY for inside China
    url_string <- paste(api_url, '?latlng=', latlng[1], ',', latlng[2], 
                        '&sensor=false', sep = '')
    if(nchar(key) > 0){
      url_string <- paste(url_string, '&key=', key, sep = '')
    }
  }
  if(api == 'baidu'){
    # coordinate type lookup table
    code <- c('wgs84ll', 'gcj02ll', 'bd09ll')
    names(code) <- c('WGS-84', 'GCJ-02', 'BD-09')
    coordtype <- code[ics]
    # http://api.map.baidu.com/geocoder/v2/?location=LAT,LNG&coordtype=COORDTYPE
    # &output=json&ak=API_KEY
    url_string <- paste(api_url, '?location=', latlng[1], ',', latlng[2], 
                        '&coordtype=', coordtype, '&output=json&ak=', key, sep = '')
  }
  
  if(messaging) message(paste('calling ', url_string, ' ... ', sep = ''), appendLF = F)
  
  # reverse gecode
  con <- curl(URLencode(url_string))
  rgc <- fromJSON(paste(readLines(con, warn = FALSE), collapse = ''))
  if(messaging) message('done.')  
  close(con)
  
  # reverse geocoding results
  if(api == 'google'){
    # did reverse geocoding fail?
    if(rgc$status != 'OK'){
      warning(paste('reverse geocode failed with status ', gc$status, ', location = "', 
                    latlng[1], ', ', latlng[2], '"', sep = ''), call. = FALSE)
      return(data.frame(address = NA))  
    }
    
    # more than one address found?
    if(length(rgc$results) > 1 && messaging){
      message(paste('more than one address found for "', latlng[1], ', ', 
                    latlng[2],  '", reverse geocoding first ... ', sep = ''), apppendLF = T)
    }
    
    rgcdf <- with(rgc$results[[1]], {data.frame(address = formatted_address, 
                                                row.names = NULL)})
    for(i in seq_along(rgc$results[[1]]$address_components)){
      rgcdf <- cbind(rgcdf, rgc$results[[1]]$address_components[[i]]$long_name)
    }
    names(rgcdf) <- c('address', sapply(rgc$results[[1]]$address_components, 
                                        function(l) l$types[1]))
  }
  if(api == 'baidu'){
    # did geocode fail?
    if(rgc$status != 0){
      warning(paste('geocode failed with status code ', rgc$status, ', location = "', 
                    latlng[1], ', ', latlng[2],  '". see more details in the response code table of Baidu Geocoding API', 
                    sep = ''), call. = FALSE)
      return(data.frame(address = NA))
    }
    
    rgcdf <- with(rgc$result, {
      data.frame(address = formatted_address, 
                 street_number = NULLtoNA(addressComponent['street_number']), 
                 street = NULLtoNA(addressComponent['street']), 
                 district = NULLtoNA(addressComponent['district']), 
                 city = NULLtoNA(addressComponent['city']), 
                 province = NULLtoNA(addressComponent['province']), 
                 row.names = NULL)})
  }
  
  if(output == 'address') return(rgcdf['address'])
  if(output == 'addressc') return(rgcdf)
}

R/geocode.R
#' Geocode
#'
#' geocodes an address using Google or Baidu Maps API. Note that in most cases by 
#' using this function you are agreeing to the Google Maps API Terms of Service 
#' at \url{https://developers.google.com/maps/terms} or the Baidu Maps API Terms 
#' of Use at \url{http://developer.baidu.com/map/law.htm}.
#' 
#' @param address a character vector specifying a location of interest (e.g., 
#' "Tsinghua Univeristy").
#' @param api use Google or Baidu Maps API. When using Baidu Maps API, the address must be in Chinese.
#' @param key an api key must be provided when calling baidu maps api. 
#' While it's unnecessary for calling google maps api.
#' @param ocs output coordinate systems including WGS-84, GCJ-02 and BD-09, which
#' are the GCSs of Google Earth, Google Map in China and Baidu Map, respectively. 
#' For address out of China, ocs is automatically set to 'WGS-84' and other values 
#' are igored.
#' @param output lat/lng coordinates or lat/lng coordinates with location type (Goolge Map) | confidence #' (Baidu Map) or lat/lng coordinates with formated address and address components (only available for #' Google Map API).
#' @param messaging turn messaging on/off. The default value is FALSE.
#' @param time the time interval to geocode, in seconds. Default value is zero. 
#' When you geocode multiple addresses, set a proper time interval to avoid 
#' exceeding usage limits. For details see 
#' \url{https://developers.google.com/maps/documentation/business/articles/usage_limits}
#' @return a data.frame with variables lat/lng or lat/lng/conf 
#' @author Jun Cai (\email{cai-j12@@mails.tsinghua.edu.cn}), PhD student from 
#' Center for Earth System Science, Tsinghua University
#' @details note that the google maps api limits to 2500 queries a day.
#' @seealso \code{\link{revgeocode}}, \code{\link{geohost}}.
#' 
#' Google Maps API at \url{http://code.google.com/apis/maps/documentation/geocoding/} 
#' and Baidu Maps API at \url{http://developer.baidu.com/map/webservice-geocoding.htm}
#' @export
#' @import plyr
#' @examples
#' \dontrun{
#' geocode('Tsinghua University', api = 'google', ocs = 'GCJ-02')
#' geocode('Tsinghua University', api = 'google', ocs = 'WGS-84', 
#'         messaging = TRUE)
#' geocode('Beijing railway station', api = 'google', ocs = 'WGS-84', 
#'         output = 'latlngc')
#' geocode('Beijing railway station', api = 'google', ocs = 'WGS-84', 
#'         output = 'latlnga')
#' geocode(c('Tsinghua University', 'Beijing railway station'), api = 'google', 
#'         ocs = 'GCJ-02')
#' geocode(c('Tsinghua University', 'Beijing railway station'), api = 'google', 
#'         ocs = 'WGS-84', output = 'latlngc', messaging = TRUE)
#' geocode(c('Tsinghua University', 'Beijing railway station'), api = 'google', 
#'         ocs = 'WGS-84', output = 'latlnga', messaging = TRUE)
#' geocode(c('Tsinghua University', 'Beijing railway station'), api = 'google', 
#'         ocs = 'WGS-84', output = 'latlngc', messaging = TRUE, time = 2)
#' geocode('Beijing railway station', api = 'baidu', key = 'your baidu maps api key', 
#' ocs = 'BD-09')
#' geocode('Beijing railway station', api = 'baidu', key = 'your baidu maps api key', 
#' ocs = 'GCJ-02', messaging = TRUE)
#' geocode('Beijing railway station', api = 'baidu', key = 'your baidu maps api key', 
#' ocs = 'BD-09', output = 'latlngc')
#' geocode(c('Tsinghua University', 'Beijing railway station'), api = 'baidu', 
#' key = 'your baidu maps api key', ocs = 'BD-09')
#' geocode(c('Tsinghua University', 'Beijing railway station'), api = 'baidu', 
#' key = 'your baidu maps api key', ocs = 'WGS-84', output = 'latlngc')
#' }

geocode <- function(address, api = c('google', 'baidu'), key = '', 
                    ocs = c('WGS-84', 'GCJ-02', 'BD-09'), 
                    output = c('latlng', 'latlngc', 'latlnga'), messaging = FALSE, 
                    time = 0){  
  # check parameters
  stopifnot(is.character(address))
  api <- match.arg(api)
  stopifnot(is.character(key))
  output <- match.arg(output)
  ocs <- match.arg(ocs)
  stopifnot(is.logical(messaging))
  stopifnot(is.numeric(time))
  
  # vectorize for many addresses
  if (length(address) > 1) {
    if (api == 'google') {
      s <- 'google restricts requests to 2500 requests a day.'
      if (length(address) > 2500) stop(s, call. = F)
      if (length(address) > 200 & messaging) message(paste('Reminder', s, sep = ' : '))
    }
    
    return(ldply(address, function(add){
      Sys.sleep(time)
      geocode(add, api = api, key = key, ocs = ocs, output = output, messaging = messaging)
    }))
  }
  
  # location encoding
  address <- enc2utf8(address)
  # different google maps api is used based user's location. If user is inside China,
  # ditu.google.cn is used; otherwise maps.google.com is used.
  if (api == 'google') {
    cname <- ip.country()
    if (cname == "CN") {
      api_url <- 'http://ditu.google.cn/maps/api/geocode/json'
    } else{
      api_url <- 'http://maps.googleapis.com/maps/api/geocode/json'
    }
  } else{
    api_url <- 'http://api.map.baidu.com/geocoder/v2/'
  }
  # format url
  # https is only supported on Windows, when R is started with the --internet2 
  # command line option. without this option, or on Mac, you will get the error 
  # "unsupported URL scheme".
  if (api == 'google') {
    # http://maps.googleapis.com/maps/api/geocode/json?address=ADDRESS&sensor
    # =false&key=API_KEY for outside China
    # http://ditu.google.cn/maps/api/geocode/json?address=ADDRESS&sensor
    # =false&key=API_KEY for outside China
    url_string <- paste(api_url, '?address=', address, '&sensor=false', sep = '')
    if (nchar(key) > 0) {
      url_string <- paste(url_string, '&key=', key, sep = '')
    }
  }
  if (api == 'baidu') {
    # http://api.map.baidu.com/geocoder/v2/?address=ADDRESS&output=json&ak=API_KEY
    url_string <- paste(api_url, '?address=', address, '&output=json&ak=', key, sep = '')
  }
  
  if (messaging) message(paste('calling ', url_string, ' ... ', sep = ''), appendLF = F)
  
  # geocode
  con <- curl(URLencode(url_string))
  gc <- fromJSON(paste(readLines(con, warn = FALSE), collapse = ''))
  if (messaging) message('done.')  
  close(con)
  
  # geocoding results
  if (api == 'google') {
    # did geocode fail?
    if (gc$status != 'OK') {
      warning(paste('geocode failed with status ', gc$status, ', location = "', 
                    address, '"', sep = ''), call. = FALSE)
      return(data.frame(lat = NA, lng = NA))  
    }
    
    # more than one location found?
    if (length(gc$results) > 1 && messaging) {
      Encoding(gc$results[[1]]$formatted_address) <- "UTF-8"
      message(paste('more than one location found for "', address, 
                    '", using address\n"', tolower(gc$results[[1]]$formatted_address), 
                    '"', sep = ''), appendLF = T)
    }
    
    gcdf <- with(gc$results[[1]], {
      data.frame(lat = NULLtoNA(geometry$location['lat']), 
                 lng = NULLtoNA(geometry$location['lng']), 
                 loctype = tolower(NULLtoNA(geometry$location_type)), 
                 address = tolower(NULLtoNA(formatted_address)),
                 row.names = NULL)})
    
    # address components
    attrdf <- ldply(gc$results[[1]]$address_components, function(l){
      l <- lapply(l, function(li) {
        if (length(li) == 0) li <- NA else li
      })
      as.data.frame(l, stringsAsFactors = FALSE)[1, ]
    })
    attrdf <- attrdf[!is.na(attrdf$types), c('types', 'long_name')]
    gcdf <- within(gcdf, {
      poi <- 
        tolower(NULLtoNA(attrdf$long_name[attrdf$types == 'point_of_interest']))
      street_no <- 
        as.numeric(NULLtoNA(attrdf$long_name[attrdf$types == 'street_number']))    
      route <- 
        tolower(NULLtoNA(attrdf$long_name[attrdf$types == 'route']))
      subloc_l3 <- 
        tolower(NULLtoNA(attrdf$long_name[attrdf$types == 'sublocality_level_3'])) # 村
      subloc_l2 <- 
        tolower(NULLtoNA(attrdf$long_name[attrdf$types == 'sublocality_level_2'])) # 镇
      subloc_l1 <- 
        tolower(NULLtoNA(attrdf$long_name[attrdf$types == 'sublocality_level_1'])) # 区|县
      locality <- 
        tolower(NULLtoNA(attrdf$long_name[attrdf$types == 'locality'])) # 地级市
      admin_area_l2 <- 
        tolower(NULLtoNA(attrdf$long_name[attrdf$types == 'administrative_area_level_2'])) # US county
      admin_area_l1 <- 
        tolower(NULLtoNA(attrdf$long_name[attrdf$types == 'administrative_area_level_1'])) # 省|直辖市 or US state
      country <- 
        tolower(NULLtoNA(attrdf$long_name[attrdf$types == 'country']))      
      postal_code <- 
        tolower(NULLtoNA(attrdf$long_name[attrdf$types == 'postal_code']))          
    })
    
    # convert coordinates only in China
    if (!outofChina(gcdf[, 'lat'], gcdf[, 'lng'])) {
      gcdf[c('lat', 'lng')] <- conv(gcdf[, 'lat'], gcdf[, 'lng'], from = 'GCJ-02', to = ocs)
    } else{
      if (ocs != 'WGS-84') {
        message('wrong usage: for address out of China, ocs can only be set to "WGS-84"', appendLF = T)
        ocs <- 'WGS-84'
      }
    }
    
    if (output == 'latlng') return(gcdf[c('lat', 'lng')])
    if (output == 'latlngc') return(gcdf[c('lat', 'lng', 'loctype')])
    if (output == 'latlnga') return(gcdf[!names(gcdf) %in% c('loctype')])
  }
  if (api == 'baidu') {
    # did geocode fail?
    if (gc$status != 0) {
      warning(paste('geocode failed with status code ', gc$status, ', location = "', 
                    address, '". see more details in the response code table of Baidu Geocoding API', 
                    sep = ''), call. = FALSE)
      return(data.frame(lat = NA, lng = NA))
    }
    
    gcdf <- with(gc$result, {data.frame(lat = NULLtoNA(location['lat']), 
                                        lng = NULLtoNA(location['lng']), 
                                        conf = NULLtoNA(confidence), 
                                        row.names = NULL)})
    
    # convert coordinates
    gcdf[c('lat', 'lng')] <- conv(gcdf[, 'lat'], gcdf[, 'lng'], from = 'BD-09', to = ocs)
    
    if (output == 'latlnga') {
      message('Baidu map geocoder cannot return address, please try Goolge map geocoder.', appendLF = T)
      output <- 'latlng'
    }
    if (output == 'latlng') return(gcdf[c('lat', 'lng')])
    if (output == 'latlngc') return(gcdf[c('lat', 'lng', 'conf')])
  }
}

# fill NULL with NA
NULLtoNA <- function(x){
  if (is.null(x)) return(NA)
  if (is.character(x) & length(x) == 0) return(NA)
  x
}

# option ip.country to store IP country and avoid calling geohost() repeatedly 
# when geocoding multiple addresses or revgeocoding multiple locations
ip.country <- function(){
  if (!"ip.country" %in% names(options())) {
    options(ip.country = geohost(api = "ipinfo.io")["country"])
  }
  getOption("ip.country")
}
